internal/host/host.go

```diff
@@ -515,6 +517,58 @@ func (m *Manager) UpdateNTP(ctx context.Context, h *models.Host, ntpSources []*m
 	return db.Model(h).Update("ntp_sources", string(bytes)).Error
 }
 
+func (m *Manager) UpdateImageStatus(ctx context.Context, h *models.Host, newImageStatus *models.ContainerImageAvailability, db *gorm.DB) error {
+	var hostImageStatuses []*models.ContainerImageAvailability
+	imageFound := false
+
+	if h.ImagesStatus != "" {
+		err := json.Unmarshal([]byte(h.ImagesStatus), &hostImageStatuses)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to unmarshal image statuses for host %s", h.ID.String())
+		}
+	}
+
+	// Check if the image status already exist
+	for index := range hostImageStatuses {
+		if hostImageStatuses[index].Name != newImageStatus.Name {
+			continue
+		}
+
+		// Same result - Nothing to update
+		if hostImageStatuses[index].Result == newImageStatus.Result {
+			return nil
+		}
+
+		// New result - Get the bigger measurments. In case of a failure the measurments should be zero.
+		newImageStatus.SizeBytes = math.Max(hostImageStatuses[index].SizeBytes, newImageStatus.SizeBytes)
+		newImageStatus.Time = math.Max(hostImageStatuses[index].Time, newImageStatus.Time)
+		newImageStatus.DownloadRate = math.Max(hostImageStatuses[index].DownloadRate, newImageStatus.DownloadRate)
+		hostImageStatuses[index] = newImageStatus
+
+		imageFound = true
+		break
+	}
+
+	if !imageFound {
+		m.log.Infof("Adding new image status for %s with status %s to host %s", newImageStatus.Name, newImageStatus.Result, h.ID.String())
+		hostImageStatuses = append(hostImageStatuses, newImageStatus)
+
+		m.metricApi.ImagePullStatus(h.ClusterID, *h.ID, newImageStatus.Name, string(newImageStatus.Result), newImageStatus.DownloadRate)
+
+		eventInfo := fmt.Sprintf("New image status %s for host %s. result: %s; time: %f seconds; size: %f bytes; download rate: %f MBps",
+			newImageStatus.Name, hostutil.GetHostnameForMsg(h), newImageStatus.Result,
+			newImageStatus.Time, newImageStatus.SizeBytes, newImageStatus.DownloadRate)
+		m.eventsHandler.AddEvent(ctx, h.ClusterID, h.ID, models.EventSeverityInfo, eventInfo, time.Now())
+	}
```

Comment:
Consider using a map for hostImageStatuses, it will make this code simpler.

Reply:
Would you mind elaborating?